### ISSUE TITLE
Antithesis: Record ingested into a a partitioned Kafka topic never becomes visible in the Kafka source, no librdkafka errors

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#7488]
### What version of Materialize are you using?

```
$ materialized v0.8.1-dev (7bb59a405)
```

### What was the issue?

In the Antithesis environment, a kafka record ingested into a multi-partition topic is never seen in the Kafka source, even after ample waiting time. Unlike in other instances, there are no librdkafka ```ERROR``` lines, so one wonders what happened to the record in question.

```
[866156] [238.475952] [workload] [info]rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 112ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 13s\n
[866156] [238.561868] [workload] [info]kafka-time-offset.td:262:1: error: non-matching rows: expected:\n
[866156] [238.561877] [workload] [info][[\hazelnut\", \"hazelnut\", \"1\"]]\n"
[866156] [238.561887] [workload] [info]got:\n
[866156] [238.561896] [workload] [info][]\n
[866156] [238.561905] [workload] [info]Diff:\n
[866156] [238.561915] [workload] [info]row missing: [\hazelnut\", \"hazelnut\", \"1\"]\n"
[866156] [238.562185] [workload] [info]\n
[866156] [238.562455] [workload] [info]     |\n
[866156] [238.562464] [workload] [info] 261 | \n
[866156] [238.56252] [workload] [info] 262 | > SELECT * FROM upsert_time_offset_5\n
[866156] [238.563069] [workload] [info]     | ^\n
[866156] [238.563833] [workload] [info]1 errors were encountered during execution\n
[866156] [238.563842] [workload] [info]files involved: kafka-time-offset.td\n
```

### Is the issue reproducible? If so, please provide reproduction instructions.

I think this is only seen in the Antithesis environment.

### Please attach any applicable log files.

[866156__kafka-time-offset_td_262_1___non-matching_rows__expected_.log.zip](https://github.com/MaterializeInc/materialize/files/6856340/866156__kafka-time-offset_td_262_1___non-matching_rows__expected_.log.zip)

**Comment by philip-stoev** [comment]
There seems to be a general problem with rows not being fetched in the face of multiple partitions. In this log:

[8135776__timestamps-kafka-avro-multi_td_91_1___non-matching_rows__expected_.log.zip](https://github.com/MaterializeInc/materialize/files/6856362/8135776__timestamps-kafka-avro-multi_td_91_1___non-matching_rows__expected_.log.zip)


We have:
1. no librdkafka error
2. ample time given for the source to catch up:
```
[8135776] [194.104678] [workload] [info]rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 112ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 12s\n
```

3. And yet some random rows are missing:
```
[8135776] [194.246893] [workload] [info]row missing: [\11\", \"1\"]\n"
[8135776] [194.246902] [workload] [info]row missing: [\7\", \"1\"]\n"
[8135776] [194.246912] [workload] [info]row missing: [\8\", \"1\"]\n"
```

4. Those rows are not even the last rows to be `kafka-ingest`-ed and they are not even ingested into the same partition.

**Comment by philip-stoev** [comment]
Here is another log about a failure in a test exactly when partitions are used:
[6774419__kafka-avro-sources_td_322_1___non-matching_rows__expected_.log.zip](https://github.com/MaterializeInc/materialize/files/6856415/6774419__kafka-avro-sources_td_322_1___non-matching_rows__expected_.log.zip)

**Comment by aljoscha** [comment]
@philip-stoev You haven't managed to repro this on a vanilla/local machine, have you?

**Comment by philip-stoev** [comment]
@aljoscha No, unfortunately this has never been seen outside of the Antithesis environment which injects both transient network issues and perturbs the timing of the individual threads.

We could add extra logging and then make sure the logging level is appropriate inside the Antithesis container and ask them to rerun their experiment.

**Comment by aljoscha** [comment]
Give me a sec, I have a suspicion that it's the metadata update frequency. By default, we only fetch new kafka partitions every 30s. With simulated network outages and messing with treads, I think it can happen that we don't see the new partition within the 60s that we give the `kafka-verify` statement. I need to fix a thing and then we can set the metadata update frequency to 10ms for that source and see what happens.

**Comment by aljoscha** [comment]
Could you let it rerun with https://github.com/MaterializeInc/materialize/pull/7556? You might want to add the same config to other tests, where you have noticed this problem. Basically, in all tests where we add partitions, which should also reduce test time a bit.

**Comment by philip-stoev** [comment]
@aljoscha all done, see PR

**Comment by aljoscha** [comment]
Should be fixed by #7556
